### ISSUE TITLE
systemd dbus api performance impact during startup

### ISSUE THREAD

**Issue by remueller** [systemd/systemd#6370]
Environment: Ubuntu 17.04, systemd 232

To test the systemd client interface I used "systemctl".
I am starting the debug.service manually during runtime and automatically at startup.

--------------------------------------------------------------------------------
#debug.service:
[Unit]
DefaultDependencies=no
After=tmp.mount local-fs.target

[Service]
Type=oneshot
ExecStart=/usr/bin/strace -o /run/startupStrace.txt /bin/systemctl start base.service
RemainAfterExit=yes

[Install]
WantedBy=sysinit.target

--------------------------------------------------------------------------------

#base.service: simulates a service which needs 3 seconds for starting
[Unit]
DefaultDependecies=no
After=tmp.mount

[Service]
ExecStart=/bin/sleep 3
Type=oneshot
RemainAfterExit=true

--------------------------------------------------------------------------------

### During Runtime:
Number of recvmsg SysCalls: 23
Received Bytes: 2696

Number of sendmsg SysCalls: 4
Send Bytes: 671

### During Startup:
Number of recvmsg SysCalls: 4270
Received Bytes: 657132

Number of sendmsg SysCalls: 4
Send Bytes: 671

-------------------------------------------------------------------------------------

You can see that during startup the performance influence is really high, more than
4000 additional syscall only on client side are necessary to execute the same functionality.
In addition there are nearly the same number of syscalls necessary for dbus and systemd itself.
The reason for that is, that systemd does not provide a client or state filtering, all changes are send
to the client, in my case systemctl and the client is responsible to filter the messages he is interested in.

Is there there any important reason why systemd does not provide a filter for client notifications?

**Comment by poettering** [comment]
I am not sure what this is supposed to show? how are those stats generated? are those the syscalls PID1 executes or the syscalls systemctl executes?

D-Bus filters in the dbus-daemon message broker. As such, PID 1 first has to send the messages to the broker before they can be filtered out there. Moreover, during early boot dbus-daemon isn't running yet, in that case communication between clients and PID 1 happen through a direct private socket, and as the broker is not involved at all then no filtering will take place at all

**Comment by remueller** [comment]
You can find the strace output from systemctl for both usecases attached.
In both usecases I am doing the same thing, starting the base.service with systemctl.

The problem from my point of view is: the client "systemctl" executes the dbus function "StartUnit" and now wants to know, when the base.service startup is finished. The only way to get these information is to subscribe to the dbus signal "JobRemoved".
In a "static" system situation this works fine, because there is only the Job "base.service" in the systemd jobqueue and therefor the client systemctl only received one Signal from "JobRemoved" when it is finished.
In high "dynamic" system situation like startup the client receives a lot of Signals "JobRemoved". The client is responsible to filter these signals to find the right job he wants to know.
This causes a lot of unnecessary extra work which results in a high CPU load.  

Maybe a solution can be, to introduce a Signal in the Unit interface like this: 

_dbus:
node /org/freedesktop/systemd1/unit/base_2eservice {
  interface org.freedesktop.systemd1.Unit {
	methods:
		...
	signals:
		ActiveStateChanged(out s activeState)
	properties:
		...._

Now the client "systemctl" consumes always the same CPU load independent of the current overall system situation. The client calls the Method "StartUnit" and subscribes to the signal "ActiveStateChanged" for the base.service and receives only one Signal when the final state is reached.

[strace-systemctl-runtime.txt](https://github.com/systemd/systemd/files/1163785/strace-systemctl-runtime.txt)
[strace-systemctl-startup.txt](https://github.com/systemd/systemd/files/1163786/strace-systemctl-startup.txt)

**Comment by paulmenzel** [comment]
Do the systemd developers agree with the analysis and the suggestion improvement?

@remueller, were able to prototype this and see if the CPU load decreases?

**Comment by remueller** [comment]
No, I haven't built a prototype yet. This is some work and a change in the DBus Interface and therefore it needs support for the systemd developers.
